### ISSUE TITLE
Chunks failing, not retrying

### ISSUE THREAD

**Issue by tremby** [TTLabs/EvaporateJS#223]
Each time I try to upload a large file I get some warnings saying that a chunk failed to upload, and the process eventually stops. If I pause and then resume it will continue.

But I'm trying to figure out why this happens. Aren't failed chunks supposed to retry automatically? Is there some config option I've missed to enable this?

My JS log has something like the following in these cases:

```
authorizedSend() upload #39 vendors.js:7070:18
getPayloadSha256Content UNSIGNED-PAYLOAD vendors.js:6969:22
CanonicalRequest PUT
/ciad-videos/development/m9DYj26YRyQqkeNA/a38e3b54-6f73-11e6-b9e1-305a3a572c5d
partNumber=39&uploadId=xmQ9s_TDnTc5xOFnH1f.scOJGV07g3q2enwTNcsYlvZvp5txV1wsoUebEQmcCc8ot9YMlObqQSDpqAFdNaKrXf7jWTJaxYPj7h.GsJhdmkEHAn.gZeLRpY1Kc5GuJJ1H
content-md5:BYOE2i5tFR7eiopMOh0+wg==
host:s3-us-west-2.amazonaws.com
x-amz-date:20160831T121732Z

content-md5;host;x-amz-date
UNSIGNED-PAYLOAD vendors.js:7270:18
makeStringToSign (V4) AWS4-HMAC-SHA256
20160831T121732Z
20160831/us-west-2/s3/aws4_request
12b90916797c462ebd89aa896fa0ec45fd9e4d1bebcb644881f505a976cbf404 vendors.js:7226:18
upload # 39 Object { method: "PUT", path: "/ciad-videos/development/m9DYj26YRy…", step: "upload #39", x_amz_headers: Object, md5_digest: "BYOE2i5tFR7eiopMOh0+wg==", contentSha256: "UNSIGNED-PAYLOAD", attempts: 1, part: Object, onErr: uploadPart/upload.onErr(), on200: uploadPart/upload.on200(), 7 more… } vendors.js:6393:26
authorizedSend got signature for step: 'upload #39'    sig: ################################################################ vendors.js:7108:32
part # 39 (bytes 239075328 -> 245366784)  reported length: 5552128 vendors.js:6371:22
getPayloadSha256Content UNSIGNED-PAYLOAD vendors.js:6969:22
  ### 6291456 vendors.js:7036:34
uploadPart 200 response for part #32     ETag: "5563c5cb409e9e08f1007cbd2ec186aa" vendors.js:6349:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2,2 // bytesLoaded: 0,6290434,5667866,3554335,1752096,0,801823 vendors.js:6875:18
  ### 6291456 vendors.js:7036:34
uploadPart 200 response for part #33     ETag: "905bca02edbeda12f23c25fb7a85b1a1" vendors.js:6349:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2 // bytesLoaded: 0,5864473,3554335,1752096,0,1031200 vendors.js:6875:18
  ### 6291456 vendors.js:7036:34
uploadPart 200 response for part #35     ETag: "62d3d8be5b8977a963eeaed58c15537e" vendors.js:6349:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2 // bytesLoaded: 0,4062238,2079775,0,1637406 vendors.js:6875:18
  ### 6291456 vendors.js:7036:34
problem uploading part #37,   http status: 0,   hasErrored: false,   part status: 2,   readyState: 4 vendors.js:6320:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2 // bytesLoaded: 0,5602334,0,2964511 vendors.js:6875:18
problem uploading part #37,   http status: 0,   hasErrored: true,   part status: 10,   readyState: 0,   isOnError vendors.js:6320:22
  ### 6291456 vendors.js:7036:34
uploadPart 200 response for part #36     ETag: "43076723ef3472365e5de0810255fe6f" vendors.js:6349:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2 // bytesLoaded: 0,0,4996128 vendors.js:6875:18
  ### 5552128 vendors.js:7036:34
uploadPart 200 response for part #39     ETag: "058384da2e6d151ede8a8a4c3a1d3ec2" vendors.js:6349:22
processPartsList()  anyPartHasErrored: false 3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2 // bytesLoaded: 0,0
```

The `anyPartHasErrored: false` seem suspect, given that a few lines before it says `problem uploading part #37 ... hasErrored: true`, but I'm not really sure what I'm looking at or looking for here.

Any ideas?

**Comment by bikeath1337** [comment]
It looks like commit ef684ff6b incorrectly removed the errored part from consideration. I will correct this and give you a branch to test, if you'd like. Thanks for the report.

**Comment by bikeath1337** [comment]
Pushed to https://github.com/TTLabs/EvaporateJS/commits/retry_issue

**Comment by fkjaekel** [comment]
I have some users complaining about the following error: `Error listing parts for getuploadparts() starting at part # 0`

Can this retry issue be related? EvaporateJS version is 1.5.2.

Thanks

**Comment by bikeath1337** [comment]
There were some retry issues in 1.5.2 and a residual issue identified today that has a fix that is waiting to be verified. 1.5.5 adds some IE compatibility and I would recommend giving it a try until we've addressed the issue from today.

**Comment by tremby** [comment]
@bikeath1337: Thanks for addressing this. I just tried again with this branch.

### Attempt 1 in Chromium:

The upload proceeded until about 90%. There are no non-200 status codes in the Chromium network panel. There are no "problem uploading" messages in the log. Yet it stopped before completion. The end of the log is like this:

```
processPartsList()  anyPartHasErrored: false 2,2,2,2,2,2,0 // bytesLoaded: 0,5308416,2719744,2801664,1835008,0
evaporate.js:1368 authorizedSend() upload #21
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:1576 CanonicalRequest PUT
/ciad-videos/development/QN93OW4bDyxqeG1o/63d7d104-6fc9-11e6-b9e1-305a3a572c5d
partNumber=21&uploadId=SPM45LsNnVJvLbjQWvmdPBN.8eBGKbtbNy_2XrdCVqq.SLsqxcg51VENJk6IFWoJbNdmhOHVdfCMxDv1g5ATGJzvlujtnN4YCuF7wxU._XkMr_cuAw74uk35Cqejd.Wi
content-md5:X2G0cQDgb0wBz25Y1iP4ZQ==
host:s3-us-west-2.amazonaws.com
x-amz-date:20160831T222256Z

content-md5;host;x-amz-date
UNSIGNED-PAYLOAD
evaporate.js:1532 makeStringToSign (V4) AWS4-HMAC-SHA256
20160831T222256Z
20160831/us-west-2/s3/aws4_request
2d2ef5f335de6ef220b009ca394258f8b5bbb54c6d9d8ce12e7e63be91d2167e
evaporate.js:697 upload # 21 Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/63d7d104…1g5ATGJzvlujtnN4YCuF7wxU._XkMr_cuAw74uk35Cqejd.Wi", step: "upload #21", x_amz_headers: Object, md5_digest: "X2G0cQDgb0wBz25Y1iP4ZQ=="…}
evaporate.js:1408 authorizedSend got signature for step: upload #21 - signature: ################################################################
evaporate.js:675 part # 21 ( bytes 125829120 -> 132120576 )  reported length: 5189197
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 15 ETag: "9c9f26f458d4e219ef6fc5ce36c80687"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2,2 // bytesLoaded: 0,3866624,3375104,2899968,0,1572864
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 19 ETag: "3699c3dcb6cde10dac7198d46500320e"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2 // bytesLoaded: 0,5013504,6291456,0,5189197
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 18 ETag: "170cbf80b4025fc4947939cb9d2777a9"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2 // bytesLoaded: 0,5586944,0,5189197
evaporate.js:1338   ### 5189197
evaporate.js:653 uploadPart 200 response for part # 21 ETag: "5f61b47100e06f4c01cf6e58d623f865"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2 // bytesLoaded: 0,5701632,0
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 17 ETag: "e3fea92ff34544170058cb82327a3c33"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2 // bytesLoaded: 0,0
evaporate.js:1217 monitorPartsProgress() Wed Aug 31 2016 15:24:23 GMT-0700 (PDT)
evaporate.js:1217 monitorPartsProgress() Wed Aug 31 2016 15:26:23 GMT-0700 (PDT)
```

A pause and resume on this upload failed, which is worse than before. It failed with "Uncaught TypeError: Illegal invocation" on evaporate.js:953.

### Attempt 2:

Seemed to go exactly the same way.

### Attempt 1 in Firefox:

Failed to even start, with a TypeError on line 953. "TypeError: 'log' called on an object that does not implement interface Console."

### Fix line 953

I fixed line 953 by changing `null` to `console`, for now, and tried again.

I tried another upload in Chromium and this time I did see a "problem uploading" warning and a 403 in the network panel: `problem uploading part #11,   http status: 403,   hasErrored: false,   part status: 2,   readyState: 4`. I don't know if this is any different because of the code change or if it's just that the network randomly failed on me this time. The 403 in the network panel is a SignatureDoesNotMatch, which seems strange given that all the others were fine.

The progress got to around 80% this time and has halted. The end of the log currently says the following:

```
upload # 17 Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #17", x_amz_headers: Object, md5_digest: "lAb6+TpToZOGuPPbPHeBiw=="…}
evaporate.js:1408 authorizedSend got signature for step: upload #17 - signature: ################################################################
evaporate.js:675 part # 17 ( bytes 100663296 -> 106954752 )  reported length: 6291456
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
s3-us-west-2.amazonaws.com/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba…smmkoSwTt7uMBcBbXaQOOC2yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD:1 PUT https://s3-us-west-2.amazonaws.com/ciad-videos/development/QN93OW4bDyxqeG1o…VZsmmkoSwTt7uMBcBbXaQOOC2yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD 403 (Forbidden)
evaporate.js:1338   ### 6291456
evaporate.js:622 problem uploading part #11,   http status: 403,   hasErrored: false,   part status: 2,   readyState: 4
evaporate.js:643 AWS Server response: code="SignatureDoesNotMatch", message="The request signature we calculated does not match the signature you provided. Check your key and signing method."upload.onErr @ evaporate.js:643xhr.onreadystatechange @ evaporate.js:1341
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 13 ETag: "b2626c4f3f9c2e6cf6e6ffc406b9fbdf"
evaporate.js:598 uploadPart # 11 - will wait 0 ms before retrying
evaporate.js:1275 setupRequest() Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #11", x_amz_headers: Object, md5_digest: "T00xlGa6yoxuQgfxp3P7VQ=="…}
evaporate.js:1368 authorizedSend() upload #11
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:1576 CanonicalRequest PUT
/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2-6fcb-11e6-b9e1-305a3a572c5d
partNumber=11&uploadId=m2ydK.M6rr9JCmxbm19zcyjPqw1VIC_WcRyxtU2FRML9DuitQkHMHTVZsmmkoSwTt7uMBcBbXaQOOC2yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD
content-md5:T00xlGa6yoxuQgfxp3P7VQ==
host:s3-us-west-2.amazonaws.com
x-amz-date:20160831T223943Z

content-md5;host;x-amz-date
UNSIGNED-PAYLOAD
evaporate.js:1532 makeStringToSign (V4) AWS4-HMAC-SHA256
20160831T223943Z
20160831/us-west-2/s3/aws4_request
1c69815a4ca9bad4a94fc30554ffadd5092544cebfe086b76d91bf1ee18e2165
evaporate.js:697 upload # 11 Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #11", x_amz_headers: Object, md5_digest: "T00xlGa6yoxuQgfxp3P7VQ=="…}
evaporate.js:1408 authorizedSend got signature for step: upload #11 - signature: ################################################################
evaporate.js:675 part # 11 ( bytes 62914560 -> 69206016 )  reported length: 6291456
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:825 part # 18 MD5 digest is NPZDGea+9HsUp/Auj8vJdA==
evaporate.js:825 part # 19 MD5 digest is 84RK1qmGFaciGO1zutFLtw==
evaporate.js:825 part # 20 MD5 digest is wex8B4fny89CsF/cBIzKGA==
evaporate.js:841 All parts have MD5 digests
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 14 ETag: "4fe7011f802a79c144220e0e2bae58aa"
evaporate.js:598 uploadPart # 18 - will wait 250 ms before retrying
evaporate.js:1275 setupRequest() Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #18", x_amz_headers: Object, md5_digest: "NPZDGea+9HsUp/Auj8vJdA=="…}
evaporate.js:598 uploadPart # 19 - will wait 250 ms before retrying
evaporate.js:1275 setupRequest() Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #19", x_amz_headers: Object, md5_digest: "84RK1qmGFaciGO1zutFLtw=="…}
evaporate.js:598 uploadPart # 20 - will wait 250 ms before retrying
evaporate.js:1275 setupRequest() Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #20", x_amz_headers: Object, md5_digest: "wex8B4fny89CsF/cBIzKGA=="…}
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2,0,0,0 // bytesLoaded: 4603904,0,0,5259264,4915200
evaporate.js:1368 authorizedSend() upload #18
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:1576 CanonicalRequest PUT
/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2-6fcb-11e6-b9e1-305a3a572c5d
partNumber=18&uploadId=m2ydK.M6rr9JCmxbm19zcyjPqw1VIC_WcRyxtU2FRML9DuitQkHMHTVZsmmkoSwTt7uMBcBbXaQOOC2yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD
content-md5:NPZDGea+9HsUp/Auj8vJdA==
host:s3-us-west-2.amazonaws.com
x-amz-date:20160831T223948Z

content-md5;host;x-amz-date
UNSIGNED-PAYLOAD
evaporate.js:1532 makeStringToSign (V4) AWS4-HMAC-SHA256
20160831T223948Z
20160831/us-west-2/s3/aws4_request
cfe572202cec25112843a4ec6ff92895f74ee5f5df86a6888f78007bfd912f0e
evaporate.js:697 upload # 18 Object {method: "PUT", path: "/ciad-videos/development/QN93OW4bDyxqeG1o/bf461ba2…yMvlxXo89Ekm9xgWJ9zx5IzT5EC15mpCAyK.InwEA.5a_T5AD", step: "upload #18", x_amz_headers: Object, md5_digest: "NPZDGea+9HsUp/Auj8vJdA=="…}
evaporate.js:1408 authorizedSend got signature for step: upload #18 - signature: ################################################################
evaporate.js:675 part # 18 ( bytes 106954752 -> 113246208 )  reported length: 6291456
evaporate.js:1271 getPayloadSha256Content UNSIGNED-PAYLOAD
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 16 ETag: "2eaa2f36a6578dd263eb2d5ad03a31b8"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2,2,2 // bytesLoaded: 6144000,0,0,6291456,1900544,0,0
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 17 ETag: "9406faf93a53a19386b8f3db3c77818b"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2,2 // bytesLoaded: 6291456,0,0,2277376,0,0
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 11 ETag: "4f4d319466baca8c6e4207f1a773fb55"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2,2 // bytesLoaded: 0,0,2670592,0,0
evaporate.js:1338   ### 6291456
evaporate.js:653 uploadPart 200 response for part # 18 ETag: "34f64319e6bef47b14a7f02e8fcbc974"
evaporate.js:1178 processPartsList()  anyPartHasErrored: false 2,2,2,2 // bytesLoaded: 0,0,0,0
evaporate.js:1217 monitorPartsProgress() Wed Aug 31 2016 15:41:15 GMT-0700 (PDT)
evaporate.js:1223 11 "not evaporating "
evaporate.js:1217 monitorPartsProgress() Wed Aug 31 2016 15:43:15 GMT-0700 (PDT)
evaporate.js:1223 11 "not evaporating "
```

It looks from that like the chunk which failed once (11) then succeeded when tried again. Yet things still seem to have halted before finishing.

Is there any other information I can gather which would help solving this?

**Comment by tremby** [comment]
Oh, and with that small logging fix, pause and resume then does cause it to continue and finish. And I then get the same behaviour in Firefox.

**Comment by bikeath1337** [comment]
Thanks for the additional info @tremby . I have to stop now but will take a look in the morning. It should be straightforward to address...or are those famous last words?

**Comment by tremby** [comment]
I looked closer at the filtered log.

In this particular attempt (the last one, where one chunk failed), the following happened:
- Parts 1 to 10 started, with all of them eventually completing successfully
- Part 11 started, and I see the request in the network tab. Its log entries at this point:

        part # 11 MD5 digest is T00xlGa6yoxuQgfxp3P7VQ==
          ### 6291456
        uploadPart # 11 - will wait 250 ms before retrying
        setupRequest() Object {…}
- Part 12 is then mentioned in the log:

        part # 12 MD5 digest is Vld29w/JsH9MOaivLBw4gg==
        evaporate.js:598 uploadPart # 12 - will wait 250 ms before retrying
        evaporate.js:1275 setupRequest() Object {...}
- Part 11 is then mentioned again: `authorizedSend()`, `getPayloadSha256Content`, `CananicalRequest`, `makeStringToSign`, and `upload # 11` lines among others, and I see its entries in the network tab. This request eventually goes on to fail, but not until...
- Then I see all the same messages about part 13, and I see its entries in the network tab. Part 12 never got this treatment, and there are no requests to sign anything nor send anything to S3 in the network tab for part 12.
- Parts 13 and 14 look fine
- Part 15 is mentioned but never asks for a signature or tries to upload, just like part 12
- Parts 16 and 17 look fine
- Then I see part 11 retrying, and this time it succeeds
- Part 18 looks fine
- Parts 19 and 20 are mentioned but don't upload, just like 12 and 15
- Then there are my logs about hitting pause and resume
- Parts 12, 15, 19, 20 then all upload with no issues.

So the question is why parts 12, 15, 19, and 20 didn't even get to the point of asking my server for a signature.

**Comment by tremby** [comment]
I dug into the code. The issue is at line 688, where the timeout is set. The timeout fires and then a test is done. In the cases I've noticed so far where a chunk is skipped, it is when `evaporatingCount < con.maxConcurrentParts` evaluates to false, i.e. there are already the maximum number of concurrent uploads happening. In this case there's no else clause -- this chunk is just forgotten about.

**Comment by tremby** [comment]
I've come up with some tweaks which make things work much more smoothly for me, in the referenced pull request.

**Comment by bikeath1337** [comment]
Thanks, @tremby . Looking at this now.

**Comment by bikeath1337** [comment]
@tremby I should have kept this line in: 02f6dd0b3. I just restored it on retry_issue. Any better? I was able to reproduce and am now seeing full downloads.

**Comment by tremby** [comment]
Yes, as I said [here](https://github.com/TTLabs/EvaporateJS/pull/226#issuecomment-244116211). I'd appreciate a rebase of this branch to remove the commits where you take the line out and then add it again, so my name's still there on the blame output. :stuck_out_tongue_winking_eye:

Here, I've [done it for you](https://github.com/tremby/EvaporateJS/commits/retry_issue); if you are agreeable, reset your branch to mine.

    git checkout retry_issue
    git fetch git@github.com:tremby/EvaporateJS.git
    git reset --hard FETCH_HEAD
    git push -f

Or do it yourself:

    git checkout retry_issue
    git rebase -i 62b783c # this is my "Set part status to pending..." commit
    # Remove the commits which remove and then replace the line, save and exit
    git push -f

I'll test shortly...

**Comment by bikeath1337** [comment]
It goes without saying that we wouldn't commit history for a WIP. Thanks again for your help here.

Though the fix to change the status "fixes" the problem as a side effect. The part's status should not be changed from PENDING until it is actually EVAPORATING. That is the real issue. 

I'm looking at this more this morning. Fortunately, with the input from the community, we'll nab this one shortly.

**Comment by bikeath1337** [comment]
I'm also seeing in-progress PUTs failing with connection reset (Chrome and Safari). Will look into it  but as the XHRs kick off, at that point, there is nothing that Evaporate can do. I am also looking at the error handling now that I've got a fairly reproducible scenario.

**Comment by tremby** [comment]
This is working well for me. Any chance of a release with this soon?

**Comment by tremby** [comment]
I might take that back... But I don't know if this is related.

I had a case just now where two parts were chronically failing. It got to the point where it was only retrying every 300 seconds and I hit pause. I then hit resume and it tried again. It got up to 100% and fired the "finished" callback, and has stayed sitting there even though more errors are coming through the console. It appears to be in an infinite loop of retrying those parts. Eventually I hit refresh and chose the same file again. It zoomed to 100% immediately, and it has stopped.

I downloaded the file again from S3 to check it, and the checksums of the files match.

**Comment by tremby** [comment]
The failures, by the way, were incorrect signatures. This is very strange, since all other signatures were just fine. What could cause this?

**Comment by bikeath1337** [comment]
This should be resoled in #236 .

**Comment by bikeath1337** [comment]
Resolved in 1.5.6, PR #239
